### PR TITLE
Load provider shared libraries relative to core runtime executable

### PR DESCRIPTION
**Pull request by RyanUnderhill**
**Description**: Uses the path of the core runtime executable/shared library to load the provider shared library.

**Motivation and Context**
- Why is this change required? What problem does it solve?
When our core code is loaded by other modules (like python), the core code is not part of the DLL search path on windows and providers will fail to load. This ensures they are loaded from the right place.

### CODE REVIEW

**Comment by jywu-msft** [comment]
Thanks! tested locally and it works.

File: onnxruntime/test/shared_lib/test_inference.cc [hunk 1]
```diff
@@ -346,11 +346,15 @@ TEST(CApiTest, DISABLED_test_custom_op_library) {
...
+  char current_directory[/*MAX_PATH*/ 260];
+  if (_getcwd(current_directory, _countof(current_directory)))
+    lib_name = current_directory;
+  lib_name += "\\custom_op_library.dll";
+
 #elif defined(__APPLE__)
   lib_name = "libcustom_op_library.dylib";
 #else
-  lib_name = "./libcustom_op_library.so";
+lib_name = "./libcustom_op_library.so";
```
Comment (jywu-msft) [inline onnxruntime/test/shared_lib/test_inference.cc:357]: If we agree that absolute paths is the way to go, we should be consistent for all platforms and document this? Can be a separate PR perhaps.

Comment (RyanUnderhill) [reply onnxruntime/test/shared_lib/test_inference.cc:357]: Yeah, I have another bug about unloading custom op libraries so I'll address it there.
